src/test/app/Offer_test.cpp

```diff
@@ -795,22 +850,831 @@ class Offer_test : public beast::unit_test::suite
+    }
+
+    void
+    testcaseCurrencyConversionEntire()
+    {
+        testcase ("Currency Conversion: Entire Offer");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const gw = Account {"gateway"};
+        auto const alice = Account {"alice"};
+        auto const bob = Account {"bob"};
+        auto const USD = gw["USD"];
+
+        env.fund (XRP (10000), gw, alice, bob);
+        env.require (owners (bob, 0));
+
+        env (trust (alice, USD (100)));
+        env (trust (bob, USD (1000)));
+
+        env.require (
+            owners (alice, 1),
+            owners (bob, 1));
+
+        env (pay (gw, alice, alice["USD"] (100)));
+        auto const bobOfferSeq = env.seq (bob);
+        env (offer (bob, USD (100), XRP (500)));
+
+        env.require (
+            owners (alice, 1),
+            owners (bob, 2));
+        auto jro = ledgerEntryOffer (env, bob, bobOfferSeq);
+        BEAST_EXPECT(
+            jro[jss::node][jss::TakerGets] == XRP (500).value ().getText ());
+        BEAST_EXPECT(
+            jro[jss::node][jss::TakerPays] == USD (100).value ().getJson (0));
+
+        env (pay (alice, alice, XRP (500)), sendmax (USD (100)));
+
+        auto jrr = ledgerEntryState (env, alice, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "0");
+        jrr = ledgerEntryRoot (env, alice);
+        BEAST_EXPECT(
+            jrr[jss::node][sfBalance.fieldName] ==
+            std::to_string(
+                XRP (10000).value ().mantissa () +
+                XRP (500).value ().mantissa () -
+                env.current ()->fees ().base * 2)
+        );
+
+        jrr = ledgerEntryState (env, bob, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-100");
+
+        jro = ledgerEntryOffer(env, bob, bobOfferSeq);
+        BEAST_EXPECT(jro[jss::error] == "entryNotFound");
+
+        env.require (
+            owners (alice, 1),
+            owners (bob, 1));
+    }
+
+    void
+    testcaseCurrencyConversionIntoDebt()
+    {
+        testcase ("Currency Conversion: Offerer Into Debt");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const alice = Account {"alice"};
+        auto const bob = Account {"bob"};
+        auto const carol = Account {"carol"};
+
+        env.fund (XRP (10000), alice, bob, carol);
+
+        env (trust (alice, carol["EUR"] (2000)));
+        env (trust (bob, alice["USD"] (100)));
+        env (trust (carol, bob["EUR"] (1000)));
+
+        auto const bobOfferSeq = env.seq (bob);
+        env (offer (bob, alice["USD"] (50), carol["EUR"] (200)),
+            ter(tecUNFUNDED_OFFER));
+
+        env (offer (alice, carol["EUR"] (200), alice["USD"] (50)));
+
+        auto jro = ledgerEntryOffer(env, bob, bobOfferSeq);
+        BEAST_EXPECT(jro[jss::error] == "entryNotFound");
+    }
+
+    void
+    testcaseCurrencyConversionInParts()
+    {
+        testcase ("Currency Conversion: In Parts");
```

Comment:
I think this is another unit test that might benefit from some form of narrative.